weave/flow/dataset.py

```diff
@@ -71,6 +75,64 @@ def to_pandas(self) -> "pd.DataFrame":
 
         return pd.DataFrame(self.rows)
 
+    def add_rows(self, rows: Iterable[dict]) -> "Dataset":
+        """Create a new dataset version by appending rows to the existing dataset.
+
+        This is useful for adding examples to large datasets without having to
+        load the entire dataset into memory.
+
+        Args:
+            rows: The rows to add to the dataset.
+
+        Returns:
+            The updated dataset.
+        """
+        client = get_weave_client()
+        if client is None:
+            raise ValueError("Weave is not initialized, call weave.init().")
+        if not isinstance(self.rows, WeaveTable) or not self.rows.table_ref:
+            raise TypeError(
+                "This dataset is not saved to weave. Call weave.publish(dataset) "
+                "to save the dataset before adding rows."
+            )
+        if self.rows.table_ref.project != client.project:
+            raise ValueError(
+                "This dataset is not saved to the same project as the current weave client. "
+                "Client is in project:  "
+                + client.project
+                + " but dataset is in project: "
+                + self.rows.table_ref.project
+            )
+        if self.rows.table_ref.entity != client.entity:
+            raise ValueError(
+                "This dataset is not saved to the same entity as the current weave client. "
+                "Client is in entity: "
+                + client.entity
+                + " but dataset is in entity: "
+                + self.rows.table_ref.entity
+            )
+        table_update_res = client._append_to_table(
+            self.rows.table_ref.digest, list(rows)
+        )
+        new_table_ref = TableRef(
+            entity=client.entity,
+            project=client.project,
+            _digest=table_update_res.digest,
+        ).uri()
+        return client._save_object(
+            {
+                # str() to prevent weave from capturing values as refs
+                "name": str(self.name),
+                "rows": new_table_ref,
+                **({"description": str(self.description)} if self.description else {}),
+                # Class metadata
+                "_type": "Dataset",
```

Comment:
Can we use client.publish? or like is there a helper? I _feel_ like we should be able to call `make_dataset` which calls `make_base_object[T] Dataset` and does the right validation + adds the private class metadata stuff. This is deep in the pydantic serialization world that @neutralino1 Tim and Jamie have been battling, I wonder if they have any ideas.

Reply:
oh client.publish doesn't exist, thats `weave.publish`

Reply:
Yeah and unfortunately the `Dataset` object cannot currently requires `rows` to be a list of dict when it gets inited. I think there is probably way to do what you are describing and not hard code the class metadata but it will require more complex changes to the class.

Reply:
Can we modify the `.publish` logic to detect whether `rows` is a list of dicts or a `TableRef`?

Let's explore what it'd take to do the right thing here.

Reply:
For posterity.
<img width="992" alt="Screenshot 2025-04-24 at 10 59 38 AM" src="https://github.com/user-attachments/assets/d265c8b9-8a16-4220-aca2-a6c0c9a6cd3c" />

Reply:
I've updated the PR so that the client helper for appending to a table now returns a properly formed `WeaveTable`. I had to make a few fields of the constructor and attributes of `WeaveTable` optional, but the uses of those fields seem to treat them as optional anyway so I don't think this is an issue. Now the dataset helper looks much cleaner and we publish using `weave.publish` directly instead of calling internal client functions from the dataset class